src/qibo_client/qibo_job.py

```diff
@@ -0,0 +1,221 @@
+
+
+class QiboJobStatus(Enum):
+    QUEUED = "to_do"
+    RUNNING = "in_progress"
+    DONE = "success"
+    ERROR = "error"
+
+
+def wait_for_response_to_get_request(
+    url: str, seconds_between_checks: T.Optional[int] = None, verbose: bool = False
+) -> T.Tuple[requests.Response, QiboJobStatus]:
+    """Wait until the server completes the computation and return the response.
+
+    :param url: the endpoint to make the request
+    :type url: str
+
+    :return: the response of the get request
+    :rtype: requests.Response
+    :return: the completed job response status
+    :rtype: QiboJobStatus
+    """
+    if seconds_between_checks is None:
+        seconds_between_checks = constants.SECONDS_BETWEEN_CHECKS
+    if not verbose:
+        logger.info("Please wait until your job is completed...")
+
+    while True:
+        response = QiboApiRequest.get(url, timeout=constants.TIMEOUT)
+        job_status = convert_str_to_job_status(response.headers["Job-Status"])
+        if verbose and job_status == QiboJobStatus.QUEUED:
+            logger.info("Job QUEUING")
+        if verbose and job_status == QiboJobStatus.RUNNING:
+            logger.info("Job RUNNING")
+        if job_status in [QiboJobStatus.DONE, QiboJobStatus.ERROR]:
+            if verbose:
+                logger.info("Job COMPLETED")
+            return response, job_status
+        time.sleep(seconds_between_checks)
+
+
+def _write_stream_to_tmp_file(stream: T.Iterable) -> Path:
+    """Write chunk of bytes to temporary file.
+
+    The tmp_path should be closed manually.
+
+    :param stream: the stream of bytes chunks to be saved on disk
+    :type stream: Iterable
+
+    :return: the name of the tempo
+
+    """
+    with tempfile.NamedTemporaryFile(delete=False) as tmp_file:
+        for chunk in stream:
+            if chunk:
+                tmp_file.write(chunk)
+        archive_path = tmp_file.name
+    return Path(archive_path)
+
+
+def _extract_archive_to_folder(source_archive: Path, destination_folder: Path):
+    with tarfile.open(source_archive, "r:gz") as archive:
+        archive.extractall(destination_folder)
+
+
+def _save_and_unpack_stream_response_to_folder(
+    stream: T.Iterable, results_folder: Path
+):
+    """Save the stream to a given folder.
+
+    Internally, save the stream to a temporary archive and extract its contents
+    to the target folder.
+
+    :param stream: the iterator containing the response content
+    :type stream: Iterable
+    :param results_folder: the local path to the results folder
+    :type results_folder: Path
+    """
+    archive_path = _write_stream_to_tmp_file(stream)
+
+    _extract_archive_to_folder(archive_path, results_folder)
+
+    # clean up temporary file
+    archive_path.unlink()
+
+
+@dataclass
+class QiboJobResult:
+    pid: str
+    success: bool
+    result: T.Optional[qibo.result.QuantumState]
+
+    def __str__(self):
+        return str(self.result)
+
```

Comment:
May I ask you to add a __call__ method, because otherwise we have to do job.result().result in order to get the result content.

Reply:
The following is the IBMQ way:

```python
job = backend.run(...)
 
try:
    job_result = job.result()  # It will block until the job finishes.
    print("The job finished with result {}".format(job_result))
except JobError as ex:
    print("Something wrong happened!: {}".format(ex))
```

Do you want to keep it and add `result = job()` syntax as well?

Reply:
The thing that seems annoying to me is:
```
r = job.result()
state = r.result.state()
```
Does IBMQ does the same?

Reply:
You should do this:

```
job.state()
r = job.result()
```

`r` is a  `QiboJobResult` instance and does not have a `state()` method:

https://github.com/qiboteam/qibo-client/blob/feature/client-does-not-hang/src/qibo_client/qibo_job.py#L105-L112

Reply:
The `QiboJobResult.success` boolean value contains whether the job was successful or not. I don't know if it's an important information that much

Reply:
Yes, indeed, IBMQ is not asking the user to call results twice, e.g.:
```
result = job.result()
print(result.state())
```
while in this PR, we have to do:
```
result = job.result()
print(result.result.state())
```
which is annoying, no?

Reply:
Ok, what is the object type you want me to return with `result.state()`?

Reply:
I think we should remap job.result() into Qibo's corresponding object, so instead of returning a `QiboJobResults`, return the content of the `result` attribute directly.

Reply:
We can drop the `QiboJobResult` class and return  the `qibo.result.QuantumState` object if you prefer

Reply:
I don't this will work because, the return object can also be a measurement, so you should return wathever qibo creates as output. You could drop this object and just keep the result quantity if this class is not used anywhere else.